### PR TITLE
if output file missing, mark task as error

### LINKED ISSUE #3287: Task reporting stuck after server rejected request
**Describe the bug**
I had a power outage. After restart BOINC started recovering and reported tasks. However server rejected request with reason "last request too recent: 568 sec". For some reason after this BOINC did not attempt to report tasks again. I found that this happened after about 12 hours later, when I noticed that almost all tasks are finished (my queue was 0.5 day) and new tasks were not downloaded. After checking logs I found part pasted below. I also saw many entries after these that WUs from this project were completed and results uploaded successfully. However WUs were not reported, I had to manually refresh project. After doing so I got bunch of new tasks. After some time BOINC triggered update again to download more tasks to fill work queue, so looks that project is not stuck anymore.

```
78	T.Brada Experiment Grid	2019-09-10 12:14:06	Sending scheduler request: To report completed tasks.	
79	T.Brada Experiment Grid	2019-09-10 12:14:06	Reporting 1 completed tasks	
80	T.Brada Experiment Grid	2019-09-10 12:14:06	Requesting new tasks for CPU	
...
83	T.Brada Experiment Grid	2019-09-10 12:14:09	Computation for task tot5_51c_SugU5Ti5sLWGHxjq53Ytpq41r_0 finished	
84	T.Brada Experiment Grid	2019-09-10 12:14:09	Output file tot5_51c_SugU5Ti5sLWGHxjq53Ytpq41r_0_r994518710_0 for task tot5_51c_SugU5Ti5sLWGHxjq53Ytpq41r_0 absent	
85	T.Brada Experiment Grid	2019-09-10 12:14:09	Starting task tot5_51c_T1snPWkhZ1NbjtUyQ66YYNsBY_0	
86	T.Brada Experiment Grid	2019-09-10 12:14:10	Computation for task tot5_51c_Ss7NaVMp95Vqp6xux8zFBSsZ1_0 finished	
87	T.Brada Experiment Grid	2019-09-10 12:14:10	Output file tot5_51c_Ss7NaVMp95Vqp6xux8zFBSsZ1_0_r1403999090_0 for task tot5_51c_Ss7NaVMp95Vqp6xux8zFBSsZ1_0 absent	
88	T.Brada Experiment Grid	2019-09-10 12:14:10	Starting task tot5_51c_SnSmCwTBbAsDs6wcPL7k8JTDk_0	
89	T.Brada Experiment Grid	2019-09-10 12:14:10	Scheduler request completed: got 0 new tasks	
90	T.Brada Experiment Grid	2019-09-10 12:14:10	Not sending work - last request too recent: 568 sec	
```

**System Information (please complete the following information):**
 - OS: CentOS 7
 - BOINC Version: 7.14.2

### PR DESCRIPTION
**Pull request by davidpanderson**
Fixes #3287

On startup, the client checks whether files marked as present
actually are present and have the right side.
If not it marks them as not present.
This is fine for downloaded files - they'll downloaded again.

For output files, this could leave the job stuck in the "uploading" state,
since there's no file to upload.

Fix this by marking the job as having an upload error in this stuation.

### CODE REVIEW

**Comment by TheAspens** [comment]
I did the following test case:

- Created a project with the uppercase sample application
- Build the client and manager from this branch
- Started the client and manager and attached to the project I built
- Waited for the task pipeline to fill with 20 or so tasks
- Suspended networking
- Let the tasks all finish (they are now in pending upload) and the slot directories are clear
- Exit the manager and client
- Go into the project directory and delete all the files that would be uploaded
- Start the client and manager again
- Re-enable networking

At this point nothing happens.  The tasks remain in "Uploading" state and the transfers remain in "Upload: pending".

I exit again and restart (just to make sure that starting up with networking disabled didn't do something) but the tasks remain in the Uploading state and the transfers remain in the "Upload: pending" state.

**Comment by davidpanderson** [comment]
That's odd - I did the same test and it works.

The function check_file_existence() (called at startup) should notice that the files
are missing, and mark the jobs as errors.
It should generate log messages saying this.

**Comment by TheAspens** [comment]
After reading the code, I think it is a transition case.  I first ran the test using a client built from master to make sure I could experience the situation.  I then built from this branch and tested to see if it cleared the issue (which it didn't).

However, that is because those files have already been marked as `fip->status == FILE_NOT_PRESENT` so it never gets into that branch of code surrounded by `if (fip->status == FILE_PRESENT) {`

Could the following new block of code:
```
            // If an output file disappears before it's uploaded,
            // flag the job as an error.
            //
            if (fip->status == FILE_NOT_PRESENT && fip->uploadable() && !fip->uploaded) {
                RESULT* rp = file_info_to_result(fip);
                if (rp) {
                    gstate.report_result_error(
                        *rp, "output file missing or invalid"
                    );
                }
            }
```

be moved outside of the `if (fip->status == FILE_PRESENT)` so that it could handle existing files in this state as well as files that are newly marked as not present?

**Comment by davidpanderson** [comment]
I made that change (though that's a zero-probability scenario)

File: client/check_state.cpp [hunk 1]
```diff
@@ -94,8 +94,9 @@ void CLIENT_STATE::check_app(APP& p) {
 }
 
 void CLIENT_STATE::check_file_info(FILE_INFO& p) {
-    if (p.pers_file_xfer) check_pers_file_xfer_pointer(p.pers_file_xfer);
-    if (p.result) check_result_pointer(p.result);
+    if (p.pers_file_xfer) {
+        check_pers_file_xfer_pointer(p.pers_file_xfer);
+    }
     check_project_pointer(p.project);
```
Comment (TheAspens) [inline client/check_state.cpp]: Last question before merge - it looks like this was a refactor to put parenthesis around the if statement.  However, it also removed the line `if (p.result) check_result_pointer(p.result);` was that intended?

**Comment by davidpanderson** [comment]
Yes - it turned out that FILE_INFO::result wasn't used or populated, and I removed it.

**Comment by TheAspens** [comment]
Great - change looks good and works.  Thanks for the change!

**Pull request merged**